libsolidity/formal/Z3CHCSmtLib2Interface.cpp

```diff
@@ -0,0 +1,214 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/formal/Z3CHCSmtLib2Interface.h>
+
+#include <libsolidity/interface/UniversalCallback.h>
+
+#include <libsmtutil/SMTLib2Parser.h>
+
+#include <boost/algorithm/string/predicate.hpp>
+
+#include <stack>
+
+using namespace solidity::frontend::smt;
+using namespace solidity::smtutil;
+
+Z3CHCSmtLib2Interface::Z3CHCSmtLib2Interface(
+	frontend::ReadCallback::Callback _smtCallback,
+	std::optional<unsigned int> _queryTimeout,
+	bool _computeInvariants
+): CHCSmtLib2Interface({}, std::move(_smtCallback), _queryTimeout), m_computeInvariants(_computeInvariants)
+{
+}
+
+void Z3CHCSmtLib2Interface::setupSmtCallback(bool _enablePreprocessing)
+{
+	if (auto* universalCallback = m_smtCallback.target<frontend::UniversalCallback>())
+		universalCallback->smtCommand().setZ3(m_queryTimeout, _enablePreprocessing, m_computeInvariants);
+}
+
+std::tuple<CheckResult, Expression, CHCSolverInterface::CexGraph> Z3CHCSmtLib2Interface::query(smtutil::Expression const& _block)
+{
+	setupSmtCallback(true);
+	std::string query = dumpQuery(_block);
+	std::string response = querySolver(query);
+	if (boost::starts_with(response, "unsat"))
+	{
+		// Repeat the query with preprocessing disabled, to get the full proof
+		setupSmtCallback(false);
+		query = "(set-option :produce-proofs true)" + query + "\n(get-proof)";
+		response = querySolver(query);
+		setupSmtCallback(true);
+		if (!boost::starts_with(response, "unsat"))
+			return {CheckResult::SATISFIABLE, Expression(true), {}};
+		return {CheckResult::SATISFIABLE, Expression(true), graphFromZ3Answer(response)};
+	}
+
+	CheckResult result;
+	if (boost::starts_with(response, "sat"))
+	{
+		auto maybeInvariants = invariantsFromSolverResponse(response);
+		return {CheckResult::UNSATISFIABLE, maybeInvariants.value_or(Expression(true)), {}};
```

Comment:
```suggestion
		return {CheckResult::SATISFIABLE, maybeInvariants.value_or(Expression(true)), {}};
```
And shouldn't here be satisfiable? Maybe I'm missing something.